### PR TITLE
feat(nx-python): added support for build with non-locked versions and non-bundle local dependencies

### LINKED ISSUE #23: The build command makes assumptions about how to include dependnecies.
- [ ] I'd be willing to implement this feature *My current time commitments within my organisation will potentially limit my involvement. I do have some time at the moment to work on this but other priorities may change things.*

## Description

I'm currently looking to use this plugin to manage internal libraries within our organisation. The structure is such that we have some base libraries that most other libraries depend on and then we have applications. We don't plan on storing the applications within the nx monorepo, only the libraries. Therefore libraries will need to be built and pushed to a pypi registry to be consumed by our applications.

The current challenge I have is that because some of the libraries depend on other libraries the current implementation of this plugin results in the whole code from the base library being included in the higher level library.

The other thing I have noticed is that version numbers get re-written with specific values. Instead of the original dependency requirements such as `=~2.3.0` being included, they get overridden with the values from the lock file.

## Motivation

This behaviour would be beneficial because I think it results in a simpler implementation that could be useful for many members of the Python community, my organisation included.

## Suggested Implementation

Ideally for our use case the implementation would be a lot simpler. Instead of copying in local dependencies they would be added as normal dependencies by looking up the package name in the `pyproject.toml` then injecting the current version from the `pyproject.toml` or using a configured version string e.g. `=~2.1.0`.

I'm not sure how to handle the current implementation. It could either be implemented as an arg / config option on the current build command or added as a new command all together. I have prototyped what would be required for the solution I suggested and it makes most of the current build command redundant.

I have created [a branch here](https://github.com/niall-mediasuite/nx-plugins/tree/simplify-build-command) which supports the changes I have addressed but at the expense of all currently functionality. I don't expect it to be used in its current state but as a tool for me to test and to provide context on the changes I'm requesting.

### PR DESCRIPTION
**Pull request by lucasvieirasilva**
## Current Behavior

the current approach uses the locked versions across all the dependencies and bundles the local dependencies in the same wheel file, like this:

`proj1` depends on a local project named `lib1` and `lib1` depends on an external package called `pendulum`.

With the current approach, the build command includes the code and brings all the external dependencies to the build package, like the following example:

```toml
[tool.poetry]
name = "pymonorepo-proj1"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  pendulum = "^2.1.2"

    [tool.poetry.dependencies.pymonorepo-lib1]
    path = "../lib1"
    develop = true
```

and

```toml
[tool.poetry]
name = "pymonorepo-lib1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_lib1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  numpy = "^1.24.1"
```

After the build the temporary `pyproject` looks this way:

```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [[tool.poetry.packages]]
  include = "pymonorepo_lib1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

    [tool.poetry.dependencies.numpy]
    version = "1.24.1 "
    markers = 'python_version >= "3.8" and python_version < "3.10"'
    optional = false

    [tool.poetry.dependencies.pendulum]
    version = "2.1.2 "
    markers = 'python_version >= "3.8" and python_version < "3.10"'
    optional = false

    [tool.poetry.dependencies.python-dateutil]
    version = "2.8.2 "
    markers = 'python_version >= "3.8" and python_version < "3.10"'
    optional = false

    [tool.poetry.dependencies.pytzdata]
    version = "2020.1 "
    markers = 'python_version >= "3.8" and python_version < "3.10"'
    optional = false

    [tool.poetry.dependencies.six]
    version = "1.16.0 "
    markers = 'python_version >= "3.8" and python_version < "3.10"'
    optional = false
```

Note, that `python-dateutil` is a dependency of `pendulum`, and the `pymonorepo_lib1` is now part of the project instead of a dependency.

## Expected Behavior

This PR implements two approaches by adding 2 options in the executor `lockedVersions` (default `true`) and `bundleLocalDependencies` (default `true`), those options can be configured in the `project.json`like this:

1. Still bundling the local dependencies together and bringing their dependencies but in the same way.

Let's consider the sample example, the temporary `pyproject` looks this way:

```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [[tool.poetry.packages]]
  include = "pymonorepo_lib1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  numpy = "^1.24.1"
  pendulum = "^2.1.2"
```

Note, the `pymonorepo_lib1` stills bundled in the project but the dependencies are listed in the same way it is on the `pyproject` (using `^`).

2. Referencing local dependencies using their versions (this behavior requires the local dependencies to be published as well)

```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  pendulum = "^2.1.2"
  pymonorepo-lib1 = "1.0.0"
```

If all the local dependency trees are going to be published to the pypi repository the second approach is better because it uses the standard pip dependency resolution.

If the local dependencies are designed to be only workspace dependencies and don't need to be published to the pypi repository the first approach is better because bundles the local dependencies.

```json
{
  "name": "lib1",
  "targets": {
    "build": {
      "executor": "@nxlv/python:build",
      "outputs": ["packages/lib1/dist"],
      "options": {
        "outputPath": "packages/lib1/dist",
        "lockedVersions": "false",
        "bundleLocalDependencies": "false"
      }
    }
  },
  "tags": []
}
```

or through CLI `npx nx run lib:build --lockedVersions=false --bundleLocalDependencies=false` (if you already configured in the project.json you don't need to provide those CLI options).

However, might have some cases in which some local dependencies need to be resolved as the standard version and some others as a bundle.

For this scenario, I'm planning to use a property in the build execution inside the `project.json` to identify if that project is marked to be published or not, if yes only reference the version, if no bundle the project.

The logic above is only required if the `bundleLocalDependencies` is set to `false`.

`packages/lib1/project.json`
```json
{
  "name": "lib1",
  "targets": {
    "build": {
      "executor": "@nxlv/python:build",
      "outputs": ["packages/lib1/dist"],
      "options": {
        "outputPath": "packages/lib1/dist",
        "publish": true,
        "lockedVersions": "false",
        "bundleLocalDependencies": "false"
      }
    }
  },
  "tags": []
}
```

For the example above if the `bundleLocalDependencies` is set to `false` the built pyproject will be in this way:

```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  pendulum = "^2.1.2"
  pymonorepo-lib1 = "1.0.0"
```

if the `publish` property is set to `false`, it will be in this way:

```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [[tool.poetry.packages]]
  include = "pymonorepo_lib1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  numpy = "^1.24.1"
  pendulum = "^2.1.2"
```

## Related Issue(s)

Fixes #23

### CODE REVIEW

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite I've published a beta version https://www.npmjs.com/package/@nxlv/python/v/15.2.0-beta.1

`@nxlv/python@15.2.0-beta.1`

Could you run some tests?

Thanks

**Comment by niall-mediasuite** [comment]
@lucasvieirasilva Sorry I totally missed you had created this PR. I'll try to give it a test later on today.

**Comment by lucasvieirasilva** [comment]
> @lucasvieirasilva Sorry I totally missed you had created this PR. I'll try to give it a test later on today.

No worries, take your time

**Comment by niall-mediasuite** [comment]
Sorry for taking so long to get back to you. So I've just tested this and here are my thoughts.

**Comment by niall-mediasuite** [comment]
Sorry for taking so long to get back to you. So I've just tested this and here are my thoughts.

First of all there is something wrong with the version you identified. I installed it but it didn't contain any of the changes so I had to build the package locally and use it from there.

After getting it going it appears to do what I would expect, in terms of the `--lockedVersions=false` and `bundleLocalDependencies=false` functionality. I haven't tested any other permutations.

On issue that I have realised is that for our use case, using a private pypi, we need to be able to set the repository for the dependency. I see there is an option to include a custom source within the `pyproject.toml` files, however we also need to tell the dependencies to use this. Would the best way to handle this just be to add a repository specifier to the build options?

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite oh ok, if you are okay with build and testing locally, I'm not going to republish the beta version,

well, regarding private repositories, I've used AWS CodeArtifact, which provides a private PyPi repo and also provides mirrors for the public PyPI, and works in this way:

1. Using a CLI you login into the private PyPi repo and the pip config files are changed to point the pip to the private repo. (because there the repo also provides a mirror feature, you can install public/private packages.
2. You can install the packages using pip without any change, however, to reference in another poetry project you must configure the custom source. (Natively, AWS CodeArtifact does not have support for Poetry, to solve this issue, I've published a CLI to handle those scenarios, here is the reference for this CLI https://pypi.org/project/aws-codeartifact-poetry/)

I don't have much experience with other private PyPi repositories, which one are you trying to use?

**Comment by niall-mediasuite** [comment]
Thanks for your reply @lucasvieirasilva. Sorry I may not have been clear in my previous post. We are using CodeArtifact too but the way we would like to configure it is to specifically tell it which packages come from our internal source. For Poetry this means specifying a source for each of our dependencies when they are included within another package.

e.g. 
```toml
[tool.poetry]
name = "pymonorepo-proj1"
version = "1.0.0"

  [[tool.poetry.source]]
  name = "foo"
  url = "https://foo.bar/simple/"

  [[tool.poetry.packages]]
  include = "pymonorepo_proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"
  pendulum = "^2.1.2"
  pymonorepo-lib1 = { version = "1.0.0", source = "foo" }
```

Does that make sense?

File: packages/nx-python/src/executors/build/executor.ts [hunk 1]
```diff
@@ -88,14 +72,33 @@ export default async function executor(
...
     );
+
+    const pythonDependency = buildTomlData.tool.poetry.dependencies.python;
+
+    buildTomlData.tool.poetry.dependencies = {};
+    buildTomlData.tool.poetry.group = {
+      dev: {
+        dependencies: {},
+      },
+    };
```
Comment (niall-mediasuite) [inline packages/nx-python/src/executors/build/executor.ts:91]: After more testing I have found this is an issue. I think all dependency groups should be handled in the same way as the main group. The only change that should be made with regards to the dependencies is to re-write the local ones to instead use the package name, or name and source, as previously discussed.

Comment (niall-mediasuite) [reply packages/nx-python/src/executors/build/executor.ts:91]: I'm new to Poetry and it looks like I've been confusing extras with groups. I'll look into extras and see if that makes sense to use instead.

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite regarding the first comment, yes that makes sense, of course not for referencing local packages in the workspace in development time, they should remain as is, however, when you build you might want to add the custom source for the local dependencies only.

In my point of view, the best approach is to parametrize in the `build` executor when the `--bundleLocalDependencies` is set to `false` the following options are available:

- `customSourceUrl`
- `customSourceName`

That could be provided in 3 ways: 

- In the `project.json` of the project (this approach allows to have multiple sources based on the project)
- In the `nx.json` by adding the `targetDefaults` property, like this:

```json
{
  ...
  "targetDefaults": {
    "build": {
      "options": {
        "lockedVersions": false,
        "bundleLocalDependencies": false,
        "customSourceUrl": "https://foo.bar/simple/",
        "customSourceName": "foo"
      }
    }
  },
  ...
}
```

This approach is going to be applied to all projects.

- In the CLI `npx nx build <myproject-name> --customSourceUrl="https://foo.bar/simple/" --customSourceName=foo`.

What do you think?

File: packages/nx-python/src/executors/build/executor.ts (hunk 1 above)
Comment (lucasvieirasilva) [reply packages/nx-python/src/executors/build/executor.ts:91]: @niall-mediasuite when poetry generates the `.tar.gz` only the main dependencies are listed in the `setup.py`, that's why in the build executor I'm only looking at the main dependencies section, 

you can prove this behavior by creating a poetry project, adding some main dependencies, adding some dependencies to a different group, and running the `poetry build` command.

Then if you look at the `setup.py` stored in the `.tar.gz` file, only the main dependencies will be listed there.

Comment (niall-mediasuite) [reply packages/nx-python/src/executors/build/executor.ts:91]: Ah, good to know. Thanks, I'll have a play around with them.

**Comment by niall-mediasuite** [comment]
@lucasvieirasilva How are you suggesting that then gets consumed? Would the `customSourceUrl` on the build parameter effect the build for the project that owns the `project.json` or would it be used by the build executor of other projects when resolving the dependency? I assume the latter, otherwise if for some reason someone had multiple different private repositories the former would be a problem.

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite in fact, forget my suggestion, currently, poetry does not compile the `source` config from `pyproject.toml` to the `setup.py`.

Look:

`pyproject.toml`
image

compiled `setup.py`
image

So, we can't do anything on the Nx plugin side to make pip identify that library needs to be downloaded from a different location.

However, as you mentioned you are using AWS CodeArtifact, if you configure to also use a mirror to public PyPi, when you log in using the AWS CLI, pip is going to download the private packages from the AWS CodeArtifact repo, and the public ones are going to mirror from the public PyPi to AWS CodeArtifact, this behavior is transparent in the pip side.

Documentation reference: https://docs.aws.amazon.com/codeartifact/latest/ug/repo-upstream-behavior.html

**Comment by niall-mediasuite** [comment]
Hi @lucasvieirasilva. Yes that's correct setup tools has no support for the source specification. However we plan on using Poetry everywhere to install dependencies. I haven't yet tested it but its my understanding Poetry would always use the `pyproject.toml` file to get dependencies, not the `setup.py`.

**Comment by niall-mediasuite** [comment]
I have since tests above when using Poetry and I can confirm that it does appear to use the dependency sources from the `pyproject.toml` to ensure the package is retrieved from the correct source.

**Comment by niall-mediasuite** [comment]
I've played around with your suggestion. See my changes [here](https://github.com/niall-mediasuite/nx-plugins/commit/22886697f25bf7773e9e66d64c772f880e617067).

Setting the source this way appears to work for us if we use Poetry to install the packages from our internal repo.

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite hey, thanks for playing and confirming this approach, I appreciate that, I looked at your changes, and it is pretty much what I was wondering to be, there is only one scenario that I noticed, if you have more than one local dependency, your changes are going to duplicate the source.

let's do this, if you are okay with my current changes, I'm gonna merge this PR and release a new version, I've raised this issue (https://github.com/lucasvieirasilva/nx-plugins/issues/40), then you push your branch and open a pull request to solve this particular issue.

Sounds like a plan?

**Comment by niall-mediasuite** [comment]
@lucasvieirasilva Yeah that sounds good. In my testing I did actually find the issue you described shortly after I just haven't pushed the change to my branch. I'll crate the PR soon.

**Comment by lucasvieirasilva** [comment]
@niall-mediasuite I've merged this PR, so, now you can raise your PR only with the changes you suggested regarding the custom source parameters.

**Pull request merged**